File: FoodOrganizer/app/src/main/java/es/fron99/foodorganize/ViewModels/ActivityTotalVM.java
package es.fron99.foodorganize.ViewModels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Objects;

import es.fron99.foodorganize.Models.Food;
import es.fron99.foodorganize.Models.Menu;
import es.fron99.foodorganize.Models.TimeMenu;

public class ActivityTotalVM extends ViewModel {

    private MutableLiveData<String> fragmentSelected;
    private MutableLiveData<ArrayList<TimeMenu>> timeMenu;
    private MutableLiveData<ArrayList<Menu>> menus;
    private MutableLiveData<ArrayList<Food>> foods;
    private Calendar daySelected;

    public ActivityTotalVM(){
        this.fragmentSelected = new MutableLiveData<>();
        this.timeMenu = new MutableLiveData<>(new ArrayList<>());
        this.menus = new MutableLiveData<>(new ArrayList<>());
        this.foods = new MutableLiveData<>(new ArrayList<>());
        daySelected = Calendar.getInstance();
    }

    public LiveData<String> getFragmentSelected(){
        if (fragmentSelected == null){
            fragmentSelected = new MutableLiveData<>();
        }
        return fragmentSelected;
    }

    public String getValueFragmentSelected(){
        if (fragmentSelected == null){
            fragmentSelected = new MutableLiveData<>();
        }
        return fragmentSelected.getValue();
    }

    public void changeFragmentSelected(String newValue){
        if (fragmentSelected == null){
            fragmentSelected = new MutableLiveData<>();
        }
        fragmentSelected.setValue(newValue);
    }

    public LiveData<ArrayList<TimeMenu>> getTimeMenus(){
        if (timeMenu == null){
            timeMenu = new MutableLiveData<>(new ArrayList<>());
        }
        return timeMenu;
    }

    public void addTimeMenu(TimeMenu newValue){
        if (timeMenu == null){
            timeMenu = new MutableLiveData<>(new ArrayList<>());
        }
        ArrayList<TimeMenu> totalTimeMenu = timeMenu.getValue();
        assert totalTimeMenu != null;
        totalTimeMenu.add(newValue);
        timeMenu.setValue(totalTimeMenu);
    }

    public void removeTimeMenu(TimeMenu newValue){
        if (timeMenu == null){
            timeMenu = new MutableLiveData<>(new ArrayList<>());
        }
        ArrayList<TimeMenu> totalTimeMenu = timeMenu.getValue();
        assert totalTimeMenu != null;
        totalTimeMenu.remove(newValue);
        timeMenu.setValue(totalTimeMenu);
    }

    public TimeMenu getTimeMenu(int i){
        if (timeMenu == null){
            timeMenu = new MutableLiveData<>(new ArrayList<>());
        }
        return timeMenu.getValue().get(i);
    }

    public LiveData<ArrayList<Menu>> getMenus(){
        if (menus == null){
            menus = new MutableLiveData<>(new ArrayList<>());
        }
        return menus;
    }

    public void addMenu(Menu newValue){
        if (menus == null){
            menus = new MutableLiveData<>(new ArrayList<>());
        }
        ArrayList<Menu> totalMenus = menus.getValue();
        assert totalMenus != null;
        totalMenus.add(newValue);
        menus.setValue(totalMenus);
    }

    public void removeMenu(Menu newValue){
        if (menus == null){
            menus = new MutableLiveData<>(new ArrayList<>());
        }
        ArrayList<Menu> totalMenu = menus.getValue();
        assert totalMenu != null;
        totalMenu.remove(newValue);
        menus.setValue(totalMenu);
    }

    public Menu getMenu(int i){
        if (menus == null){
            menus = new MutableLiveData<>(new ArrayList<>());
        }
        return menus.getValue().get(i);
    }

    public LiveData<ArrayList<Food>> getFoods(){
        if (foods == null){
            foods = new MutableLiveData<>();
        }
        return foods;
    }

    public void addFood(Food newValue){
        if (foods == null){
            foods = new MutableLiveData<>(new ArrayList<>());
        }
        ArrayList<Food> totalFoods = foods.getValue();
        assert totalFoods != null;
        totalFoods.add(newValue);
        foods.setValue(totalFoods);
    }

    public void removeFood(Food newValue){
        if (foods == null){
            foods = new MutableLiveData<>(new ArrayList<>());
        }
        ArrayList<Food> totalFood = foods.getValue();
        assert totalFood != null;
        totalFood.remove(newValue);
        foods.setValue(totalFood);
    }

    public Food getFood(int i){
        if (foods == null){
            foods = new MutableLiveData<>(new ArrayList<>());
        }
        return Objects.requireNonNull(foods.getValue()).get(i);
    }

    public Calendar getDaySelected(){
        return (Calendar) daySelected.clone();
    }

    public void setDaySelected(Calendar calendar){
        daySelected = calendar;
    }

}

File: FoodOrganizer/app/src/main/java/es/fron99/Foodorganize/ViewModels/ActivityTotalVM.kt
package es.fron99.Foodorganize.ViewModels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import es.fron99.Foodorganize.Dao.Model.*
import es.fron99.Foodorganize.Repository.Repository
import java.util.*

@Suppress("unused")
class ActivityTotalVM(application: Application) : AndroidViewModel(application) {

    /********************************************************FragmentSelected********************************************************/

    private var fragmentSelected: MutableLiveData<String> = MutableLiveData("FragmentCalendarMenus")

    fun getFragmentSelected() : LiveData<String>{
        return fragmentSelected
    }

    fun changeFragmentSelected(newValue: String) {
        this.fragmentSelected.value = newValue
    }

    /********************************************************ActivitySelected********************************************************/

    private var activitySelected: MutableLiveData<String> = MutableLiveData("Init")

    fun getActivitySelected() : LiveData<String>{
        return activitySelected
    }

    fun changeActivitySelected(newValue: String) {
        this.activitySelected.value = newValue
    }

    /********************************************************DaySelected********************************************************/

    private var daySelected: MutableLiveData<Calendar> = MutableLiveData(Calendar.getInstance())

    fun getDaySelected() : LiveData<Calendar> {
        return this.daySelected
    }

    fun changeDaySelected(date: Calendar) {
        this.daySelected.value = date
    }

    fun getValueDaySelected() : Calendar? {
        return this.daySelected.value
    }

    /********************************************************ObjSelected********************************************************/

    var timeMenuSelected: TimeMenuWithMenus = TimeMenuWithMenus()
    var menusSelected: MenuWithFoods = MenuWithFoods()
    var foodsSelected: FoodDao = FoodDao()

    /********************************************************Properties********************************************************/

    var menus: LiveData<List<MenuWithFoods>>
    var foods: LiveData<List<FoodDao>>
    fun timeMenus(): LiveData<List<TimeMenuWithMenus>> {
        return Transformations.switchMap( daySelected as LiveData<*>) {
            daySelected.value?.let { repository.getTimeMenusByDate(getApplication(), it) } }
    }

    private var repository : Repository = Repository()

    /********************************************************Init********************************************************/

    init {
        //TODO Cambiar a llamadas en hilos
        menus = repository.getMenus(application)
        foods = repository.getFoods(application)
    }

    /*************************************************************Food***********************************************************/

    fun insertFood(food: FoodDao){
        repository.insertFoods(getApplication(), food)
    }

    fun updateFood(food: FoodDao){
        repository.updateFoods(getApplication(), food)
    }

    fun dropFood(food: FoodDao){
        repository.deleteFoods(getApplication(), food)
    }

    /*************************************************************Menu***********************************************************/
/*
    fun insertMenu(menu : MenuWithFoods){
        repository.insertMenus(getApplication(), menu)
    }



    fun updateMenu(menu : MenuWithFoods){
        repository.updateMenus(getApplication(), menu)
    }

 */

    fun dropMenu(menu: MenuDao){
        repository.deleteMenus(getApplication(), menu)
    }


    /*************************************************************TimeMenu***********************************************************/

    /*
    fun insertTimeMenu(timeMenu : TimeMenuWithMenus){
        repository.insertTimeMenus(getApplication(), timeMenu)
    }

    fun updateTimeMenu(timeMenu : TimeMenuWithMenus){
        repository.updateTimeMenus(getApplication(), timeMenu)
    }

     */

    fun dropTimeMenu(timeMenu: TimeMenuDao){
        repository.deleteTimeMenus(getApplication(), timeMenu)
    }

}